### ISSUE TITLE
implementation of __hash__ method in Position class should be revised

### ISSUE THREAD

**Issue by vitalije** [leo-editor/leo-editor#535]
I believe that current `__hash__` method implementation in leoNodes.Position class is wrong. Above the implementation there is a comment line:
```python
# This has makes positions hashable, at long long last.

def __hash__(self):
    return sum([z[1] for z in self.stack])
```
I am not sure what comment means and why hashing of positions was needed nor where it is used, but Python documentation is explicit about implementing `__hash__` functions that they should not be implemented on mutable objects. If used as a key in dictionary, position can be later mutated and its hash will change so it may be in wrong slot and therefore impossible to find.

There are two possibilities: 

1. make Position class immutable
2. let `Position.__hash__` be `None`

**Comment by edreamleo** [comment]
The Position class can not be made immutable because of the p.moveToX methods.

Here is what the reference docs say about [__hash__](https://docs.python.org/3/reference/datamodel.html#object.__hash__):  "Called by built-in function hash() and for operations on members of hashed collections including set, frozenset, and dict. __hash__() should return an integer. The only required property is that objects which compare equal have the same hash value."

Imo, the present p.__hash__ function meets that criterion *at the moment that the hash is computed*. Yes, the hash will become invalid if p changes. That may well limit the usefulness of the hash.

But the problem is much deeper that hashes.  *All* positions potentially become invalid whenever *any part* of the outline changes.  That's why c.deletePositionsInList is so tricky.

In short, p.__hash__ is "formally" correct, but is useful only when the outline doesn't change. Declaring that positions aren't hashable will merely obscure the underlying problem. Devs and script writers must be aware of the subtleties and use p.hash() only as appropriate.

**Comment by vitalije** [comment]
The same page from reference documentation just a bit further:

> If a class defines mutable objects and implements an `__eq__()` method, it should not implement `__hash__()`, since the implementation of hashable collections requires that a key’s hash value is immutable (if the object’s hash value changes, it will be in the wrong hash bucket).

But I won't argue about it.

**Comment by edreamleo** [comment]
> If a class defines mutable objects and implements an `__eq__()` method, it should not implement `__hash__()`, since the implementation of hashable collections requires that a key’s hash value is immutable (if the object’s hash value changes, it will be in the wrong hash bucket).

I think the distinction we are discussing is between "potentially mutable" and "actually mutated".  All positions are potentially mutable.  It seems to me that it should be safe to enter Positions into dicts etc., if and only if they won't actually change.  In other words, `__eq__` and `__hash__` can coexist in some situations, contrary to the above quote.  But perhaps I am mistaken about this.

In any case, I don't see how eliminating `p.__hash__` is going to make Leo more reliable. It would surely break existing code that might actually work ;-)

**Comment by vitalije** [comment]
What I meant is that allowing position to be hashable, makes possible subtle bugs that are hard to spot and fix. If position was not hashable, Python interpreter would caught such possible bugs and prevent them. Let  us consider small example:

```python
a = set()
p1 = c.p
if not p1.hasChildren():
    p2 = p1.insertAsLastChild()
    p2.h = 'dummy'
    c.redraw()
else:
    p2 = p1.threadNext()
# now p1 and p2 have different hash values
h1 = hash(p1)
h2 = hash(p2)
assert h1 != h2, 'selected node must have children'

a.add(p1)
g.es('p1 in a' if p1 in a else 'p1 not in a')
g.es('p2 in a' if p2 in a else 'p2 not in a')
g.es('p1 == p2' if p1 == p2 else 'p1 != p2')
g.es('moving p1, and now')
p1.moveToThreadNext()
g.es('p1 == p2' if p1 == p2 else 'p1 != p2') # expect that p1 == p2 now
g.es('p1 in a' if p1 in a else 'p1 not in a')
g.es('a has',len(a),'element(s)')
g.es('p1 in list(a)' if p1 in list(a) else 'p1 not in list(a)')
import leo.core.leoNodes as leoNodes
f = leoNodes.Position.__hash__
try:
    leoNodes.Position.__hash__ = None
    b = set()
    b.add(p1)
except Exception as e:
    g.es_exception(e)
finally:
    leoNodes.Position.__hash__ = f
g.es(hash(p1)) # let us check that original __hash__ method has been restored
```

Making position hashable gives programmer false impression that it is safe to make a set of positions or use them as keys in dict. However it can lead to hard to debug situations because it breaks other contracts of other objects like sets and dicts. For a set we expect that it can always safely respond if it contains or not some object. It is not sane to have a set which would claim it doesn't contain p1, but if we use iterator over it  we can get p1 out of it.

Setting `Position.__hash__` to None, will make positions unhashable and interpreter would protest loudly if anyone tries to put them in set or dict.

Question of position validity after making tree changes is IMO orthogonal to this problem. Suppose for a moment that Position is immutable. Now, even if a position becomes invalid because of changes in tree, 
all sets and dictionaries that have among members or keys that position, they still would be able to give correct response on querying if that (now invalid) position is member or not and what value was associated to it. Position may become invalid, but sets and dictionaries would work as expected. User probably wouldn't be able to do much with invalid position, but it is another problem.

**Comment by edreamleo** [comment]
Many thanks for this excellent discussion.  I am convinced that making Position objects hashable is a bad idea.  What happens to Leo's unit tests when `p.__hash__` returns None?

**Comment by vitalije** [comment]
I have just run unitTest.leo with `Position.__hash__ = None`  and the only one test that is failing is using `aList2 = list(set(aList))` to check that `aList` contains unique positions. That test can be fixed by manually checking elements of list. I don't know if there is any other usage of Position hash-ability in Leo's code base.

**Comment by edreamleo** [comment]
@vitalije How should we resolve this issue?  Is a new branch a good idea, or should we just set `Position.__hash__ = None` in master?

**Comment by vitalije** [comment]
I have put in my copy of Leo `Position.__hash__ = None`. Quicksearch plugin was relying on position being hashable and in my own copy I have fixed it so that it doesn't rely on this. Apart from this I haven't noticed any other problem. Maybe there will be problems with some other plugin, but I believe they also can be fixed without too much trouble. All unit tests pass with new code (except the one I have mention before, but it fails only because test itself depends on hashablility of position class).

If you agree, I can push new version with fixed quicksearch plugin.

**Comment by edreamleo** [comment]
On Fri, Aug 11, 2017 at 4:10 AM, Виталије Милошевић <
notifications@github.com> wrote:

> I have put in my copy of Leo Position.__hash__ = None. Quicksearch plugin
> was relying on position being hashable and in my own copy I have fixed it
> so that it doesn't rely on this. Apart from this I haven't noticed any
> other problem. Maybe there will be problems with some other plugin, but I
> believe they also can be fixed without too much trouble. All unit tests
> pass with new code (except the one I have mention before, but it fails only
> because test itself depends on hashablility of position class).
>
> If you agree, I can push new version with fixed quicksearch plugin.
>
​Please do.  Many thanks for this work.

Edward

**Comment by edreamleo** [comment]
Rev a738571 fixes a crasher in the clone-find commands caused by newly-unhashable positions.

I was happy to make the required change, but it looks like there is no real gain in reliability. That is, the new code simulates a set this way:
```python
# old
clones.add(p.copy())
# new
if not p in clones:
    clones.append(p.copy())
```
Here, the *list* will be valid if and only if the *set* is valid.

It's possible that making Positions unhashable may help somebody understand the issues, but I doubt it ;-) Otoh, I see no reason to reopen this item.

One more thing.  The code above depends on the "in" operator of the Position class.  I tested that as follows:

```python
aList = [p]
print(p in aList)
print(p.copy() in aList)
```
As expected, both print statements print "True".

**Comment by vitalije** [comment]
> Here, the list will be valid if and only if the set is valid.

In this case, yes. And in many other cases, this is true. But there are cases when list will be valid and set won't. Making Position unhashable is meant to prevent those cases. We loose some speed in lookups, but in return we gain protection from nasty bugs in some special cases.

**Comment by edreamleo** [comment]
On Mon, Aug 14, 2017 at 1:27 PM, Виталије Милошевић <
notifications@github.com> wrote:

> Here, the list will be valid if and only if the set is valid.
>
> In this case, yes. And in many other cases, this is true. But there are
> cases when list will be valid and set won't. Making Position unhashable is
> meant to prevent those cases. We loose some speed in lookups, but in return
> we gain protection from nasty bugs in some special cases.
>
​I agree. That's the reason I left this issue closed.

Edward
